graphql_compiler/macros/macro_expansion.py

```diff
@@ -28,21 +29,144 @@ def _merge_non_overlapping_dicts(merge_target, new_data):
+            if ast.selection_set is not None or len(ast.directives) > 0
+        }
+
+    # Get a deterministic ordering of the merged selections
+    selection_name_order = [
+        ast.name.value
+        for ast in selection_set_a.selections
+        if ast.name.value not in selection_dict_b
+    ] + [
+        ast.name.value
+        for ast in selection_set_b.selections
+    ]
+
+    # Make sure that all property fields come before all vertex fields.
+    return SelectionSet([
+        merged_selection_dict[name]
+        for name in selection_name_order
+        if name in merged_selection_dict and merged_selection_dict[name].selection_set is None
+    ] + [
+        merged_selection_dict[name]
+        for name in selection_name_order
+        if name in merged_selection_dict and merged_selection_dict[name].selection_set is not None
+    ])
+
+
+def _merge_selection_into_target(target_ast, selection_ast, subclass_sets=None):
+    """Add the selections, directives, and coercions from the selection_ast to the target_ast."""
+    # Deal with coercions at the target
+    continuation_ast = selection_ast
+    first_selection = selection_ast.selection_set.selections[0]
+    if isinstance(first_selection, InlineFragment):
+        coercion_class = first_selection.type_condition.name.value
+        if isinstance(target_ast, InlineFragment):
+            target_class = target_ast.type_condition.name.value
+            continuation_ast = first_selection
+            target_ast.type_condition = first_selection.type_condition
+
+            # coercion_class is required to subclass target_class so we can merge the type coercions
+            if coercion_class != target_class:
+                if subclass_sets is None:
+                    raise AssertionError(u'Cannot prove type coercion at macro target is valid. '
+                                         u'Please provide a proof that {} subclasses {} using the '
+                                         u'subclass_sets argument.'
+                                         .format(coercion_class, target_class))
+                else:
+                    if (target_class not in subclass_sets or
+                            coercion_class not in subclass_sets[target_class]):
+                        raise GraphQLCompilationError(u'Invalid type coercion at macro target. {} '
+                                                      u'is expected to subclass {}.'
+                                                      .format(coercion_class, target_class))
+        else:
+            # TODO(bojanserafimov): When compiling the macro, compute the type at the
+            #                       target and record that in the macro descriptor.
+            raise NotImplementedError(u'Cannot coerce from macro that does not end '
+                                      u'in a coercion.')
+
+    # Merge the continuation into the target
+    target_ast.directives += continuation_ast.directives
```

Comment:
We are mutating a function parameter here. Let's not do this, unless there is very good reason to.